Initial green thread prototype

davidwrighton opened the pull request:
Initial stab at enabling green threads.
- Implement via the strategy of stack segments, with a greenfield implementation of concept to avoid IP concerns although the design is not that unusual
  - This is likely not a viable long term strategy, but it should work for the purposes of a prototype
- Tie it into Tasks. An infinite wait in a Green Thread will suspend the GC thread. Waits with timeouts are not integrated.

Issues...
- While the suspension resumption logic is capable of working with multiple segments, our GC stackwalker is not yet, and the JIT does not generate the appropriate prolog, so the size of a stack segment is very large.
- GC tracking in suspended threads is not enabled. If a GC occurs, there will be heap corruption
- ThreadStatics are not integrated
- PInvokes to non-green-thread safe apis do not trigger thread pinning, and so can result in incorrect behavior
- Unique managed thread objects/ids are not implemented
- JIT does not generate the new prolog
- Integration into `Task.Wait` is probably wrong in various ways.
- FCalls do not swap to the OS stack
- PInvokes do not swap to the OS stack

davidwrighton committed:
Initial jump into green thread works

davidwrighton committed:
Everything but resumption seems to work

davidwrighton committed:
Ok, it seems to work now for both suspension and resumption.

davidwrighton committed:
Phase 1 ... but it doesn't work yet. Yield needs to be handled differently

davidwrighton committed:
Miraculously, it actually appears to work

AntonLapounov commented on src/coreclr/System.Private.CoreLib/src/System/Threading/GreenThreadTasks.CoreCLR.cs:
```diff
@@ -0,0 +1,111 @@
+            }
+        }
+
+        static partial void RunOnActualGreenThread(Action action, ref bool ranAsActualGreenThread)
```
Is there a reason to use `ref bool` instead of `out bool`?

davidwrighton replied:
Yes, partial methods are not permitted to have out parameters. Effectively, if the method implementation does not exist, then the method isn't called and any out parameters would have no meaning.

AntonLapounov commented on src/coreclr/vm/amd64/AsmHelpers.asm:
```diff
@@ -698,5 +698,240 @@ NESTED_END OnCallCountThresholdReachedStub, _TEXT
+
+
+; extern "C" void YieldOutOfGreenThreadHelper(StackRange *pOSStackRange, uint8_t* osStackCurrent, uint8_t** greenThreadStackCurrent);
+; This function is supposed to hop back to the OS thread, with the original OS thread stack in place, and 
```
The comment seems unfinished.

davidwrighton replied:
It does doesn't it. :) Thank you for reading enough to find this. I'll see if I can remember the thought I had before I went and got distracted trying to put something else together.

AntonLapounov commented on src/coreclr/vm/amd64/AsmHelpers.asm:
```diff
@@ -698,5 +698,240 @@ NESTED_END OnCallCountThresholdReachedStub, _TEXT
+
+extern GetResumptionStackPointerAndSaveOSStackPointer:proc
+
+; This is a paired function with the _more_stack function to allow for 
```
And this one.

AntonLapounov commented on src/coreclr/vm/threads.h:
```diff
@@ -1245,11 +1245,7 @@ class Thread
-        LIMITED_METHOD_CONTRACT;
-        return (m_State & TS_ExecutingOnAltStack);
-    }
+    DWORD IsExecutingOnAltStack();
```
I am wondering why the implementation was moved to .cpp.

davidwrighton replied:
Ah, I should move it back. Initially I had this checking the thread local variables to report this, but I realized that we use during GC stack walking, and checking the current threads green thread state is a bad idea.

davidwrighton committed:
Address Code Review comments
- Add comments
- Remove unnecessary changes
- put necessary changes under ifdef if they are not clearly green thread related

kouvel commented on src/coreclr/vm/amd64/AsmHelpers.asm:
```diff
@@ -698,5 +698,246 @@ NESTED_END OnCallCountThresholdReachedStub, _TEXT
+
+        ; Swap to new stack
+        mov             rbx, rsp ; Stash old stack pointer into rbx (saved register)
+        mov             rsp, rax
```
Probably ok for now, but wondering if it would be feasible to use rsp for the virtual stack pointer at all, as I figure the OS could interrupt the thread and end up using the virtual stack while it's close to the limit.

davidwrighton replied:
Agreed. We've had an interesting idea around this. At the moment I leave a fair bit of space as a "red-zone" but I'm not convinced it is a good idea long-term. Send me a message on Teams to discuss whenever you'd like to.

davidwrighton committed:
Add appropriate build magic and ifdefs so that non-Windows non-x64 builds actually work
